File: game/scripts/vscripts/pve/abilities/heroes/centaur/centaur_short_attack/centaur_short_attack.lua
centaur_short_attack = class({})
function centaur_short_attack:GetCastAnimationCustom()		return ACT_DOTA_CAST_ABILITY_1 end 
function centaur_short_attack:GetPlaybackRateOverride()    return   0.8 end
function centaur_short_attack:GetCastPointSpeed() 			return 0 end

function centaur_short_attack:OnAbilityPhaseStart()
    local caster = self:GetCaster()
    local origin = caster:GetAbsOrigin()
    self.radius = self:GetSpecialValueFor("radius")
    self.radius_marker_modifier = CreateTimedRadiusMarker(caster, origin, self.radius, self:GetCastPoint(), 0.2, RADIUS_SCOPE_PUBLIC):FindModifierByName('radius_marker_thinker')

    return true
end

function centaur_short_attack:OnSpellStart()
    local caster = self:GetCaster()
    local origin = caster:GetAbsOrigin()
    local stun_duration = 0.5
    local damage_table = {
        attacker = caster,
        damage = 20,
        damage_type = DAMAGE_TYPE_PURE,
    }

    EmitSoundOn('Hero_Centaur.HoofStomp', caster)

    ApplyCallbackForUnitsInArea(caster, origin, self.radius, DOTA_UNIT_TARGET_TEAM_ENEMY, function(unit)
        damage_table.victim = unit
        ApplyDamage(damage_table)
        unit:AddNewModifier(caster, self, 'modifier_generic_stunned', { duration = stun_duration })
    end)

    EFX("particles/econ/items/centaur/centaur_ti6/centaur_ti6_warstomp.vpcf", PATTACH_WORLDORIGIN, nil, {
        cp0 = origin,
        cp1 = Vector(self.radius, 0, 0), 
        release = true
    })
end

function centaur_short_attack:OnAbilityPhaseInterrupted()
end

if IsClient() then require("wrappers/abilities") end
Abilities.Castpoint(centaur_short_attack)

File: src/vscripts/pve/abilities/heroes/centaur/centaur_short_attack.ts
import { CustomAbility } from "../../../../abilities/framework/custom_ability";
import { registerAbility } from "../../../../lib/dota_ts_adapter";
import { ModifierStun } from "../../../../modifiers/modifier_stunned";
import { precache, resource } from "../../../../precache";
import { createTimedRadiusMarker, findUnitsInRadius } from "../../../../util";

const resources = precache({
    cast: resource.fx("particles/econ/items/centaur/centaur_ti6/centaur_ti6_warstomp.vpcf")
});

@registerAbility("centaur_short_attack")
class CentaurShortAttack extends CustomAbility {
    marker?: ReturnType<typeof createTimedRadiusMarker>;

    GetAnimation() {
        return GameActivity.DOTA_CAST_ABILITY_1;
    }

    GetPlaybackRateOverride() {
        return 0.8;
    }

    GetCastingCrawl() {
        return 0;
    }

    OnAbilityPhaseStart() {
        if (super.OnAbilityPhaseStart()) {
            this.marker = createTimedRadiusMarker(
                this.caster,
                this.caster.GetAbsOrigin(),
                this.GetSpecialValueFor("radius"),
                this.GetCastPoint(),
                0.2,
                "public"
            );

            return true;
        }
        return false;
    }

    OnSpellStart() {
        const origin = this.caster.GetAbsOrigin();
        const radius = this.GetSpecialValueFor("radius");

        EmitSoundOn("Hero_Centaur.HoofStomp", this.caster);

        const enemies = findUnitsInRadius(
            this.caster,
            origin,
            radius,
            UnitTargetTeam.ENEMY,
            UnitTargetType.HERO + UnitTargetType.BASIC,
            UnitTargetFlags.NONE,
            FindOrder.ANY
        );

        for (const enemy of enemies) {
            ApplyDamage({
                victim: enemy,
                attacker: this.caster,
                damage: 20,
                damage_type: DamageTypes.PURE
            });

            ModifierStun.apply(enemy, this.caster, undefined, { duration: 0.5 });
        }

        EFX(resources.cast.path, ParticleAttachment.WORLDORIGIN, undefined, {
            cp0: origin,
            cp1: Vector(radius, 0, 0),
            release: true
        });
    }
}
